my-sms/db/migrate/20210208041633_create_titles.rb

```diff
@@ -0,0 +1,70 @@
+class CreateTitles < ActiveRecord::Migration
+  def up
+    available_titles = %w[Mr Mrs Miss Ms Master]
+
+    # do not perform migration if any unexpected titles found in Students
+    # may be more efficient to do via SQL query?
+    Student.find_in_batches do |batch|
+      batch.uniq{|s| s[:title]}.each do |s|
+        if available_titles.exclude? s.title
+          raise "Unexpected Student title #{s.title} in "\
+          "Student id #{s.id}. Migration expects titles "\
+          "in Students column to be one of #{available_titles}."
+        end
+      end
+    end
+
+    # create titles table
+    create_table :titles do |t|
+      t.string :name
+
+      t.timestamps
+    end
+
+    # pre-populate titles with english honorifics
+    available_titles.each do |name|
+      Title.create({ name: name })
+    end
+
+    # cache titles from db to avoid multiple queries
+    title_name_to_id = {}
+    Title.all.each do |title|
+      title_name_to_id[title.name] = title.id
+    end
+
+    # update students table
+    change_table :students do |t|
+      t.references :title
+    end
+
+    # add title model association
+    # without making assumptions about model
+    Student.reset_column_information
+    Student.find_each do |student|
+      student.update_attribute(:title_id, title_name_to_id.fetch(student.title, nil))
+      student.save
+    end
+
+    # remove old title (string) column
+    remove_column :students, :title
+  end
```

Comment:
Another approach (in pseudo-code) could be to:

- create `titles` table
- rename `title` to `title_old` in `students` table
- for each `student` in `students`: create a new title if it doesn't exist and set that as `student.title` (you can use `title.find_or_create_by(name: ...)`)
- remove `title_old` column in `students` table

Reply:
Ended up implementing something like this in addition to Xavier's comments in 453cd248af4c2a2ecca026d7bcae849771f6fa83
Works out much cleaner thanks.